docs/logging.md

```diff
@@ -22,3 +77,92 @@ code, time to process, (optional perf data details, e.g. database time/queries,
+great deal of Python processing overhead to marshall the data from the
+database into Django objects that is not accounted for in these
+numbers.
+
+## Blueslip frontend error reporting
+
+We have a custom library, called `blueslip` (named after the form used
+at MIT to report problems with the facilities), that takes care of
+reporting JavaScript errors.  In production, this means emailing the
+server administrators (though the setting controlling this,
+`BROWSER_ERROR_REPORTING`, is disabled by default, since JavaScript
+errors tend to have fairly high volume and are almost never
+configuration problems).  In development, this means displaying a
+highly visible overlay over the message view area, to make exceptions
+in testing a new feature hard to miss.
+
+* Blueslip is implemented in `static/js/blueslip.js`.
+* In order to capture essentially any error occuring in the browser,
+blueslip does the following:
+  * Wraps every function passed into `$.ready()`, i.e., every
+  on-webapp-startup method used by Zulip.
+  * Wraps every jQuery AJAX request handler used by Zulip.
+  * Wraps every function passed into `$.on()`, i.e. all event
+  handlers declared in Zulip.
+  * Declares a default browser exception handler.
+  * Has methods for being manually triggered by Zulip JavaScript code
+    for warnings and assertion failures.
+* Blueslip keeps a lot of all the notices it has received during a
+  browser session, and includes them in reports to the server, so that
+  one can see cases where exceptions chained together.  You can print
+  this log from the browser console using `blueslip.get_log()`.
+
+Blueslip supports several error levels:
+* `blueslip.fatal`: For fatal errors that cannot be easily recovered
+  from.  We try to avoid using it, since it kills the current JS
+  thread, rather than returning execution to the caller.  Unhandled
+  exceptions in our JS code are treated like `blueslip.fatal`.
+* `blueslip.error`: For logging of events that are definitely caused
+  by a bug and thus sufficiently important to be reported, but where
+  we can handle the error without creating major user-facing problems
+  (e.g. an exception when handling a presence update).
+* `blueslip.warn`: For logging of events that are a problem but not
+  important enough to send an email about in production.  They are,
+  however, highlighted in the JS console in development.
+* `blueslip.log` (and `blueslip.info`): Logged to the JS console
+  development and also in the blueslip log in production.  Useful for
+  data that might help discern what state the browser was in during an
+  error (e.g. whether the user was in a narrow).
+* `blueslip.debug`: Similar to `blueslip.log`, but are not printed to
+  the JS console in development.
+
+## Frontend performance reporting
+
+In order to make it easier to debug potential performance problems in
+the critically latency-intensive message sending code pathway, we log
```

Comment:
"latency-sensitive"? "Latency-intensive" sounds like it should mean that it has a lot of latency.